src/NodaTime/TimeInterval.cs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2009 The Noda Time Authors. All rights reserved.
+// Use of this source code is governed by the Apache License 2.0,
+// as found in the LICENSE.txt file.
+
+using JetBrains.Annotations;
+using NodaTime.Text;
+using System;
+using NodaTime.Utility;
+
+namespace NodaTime
+{
+    /// <summary>
+    /// An interval between two local time values (start and end).
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The TimeInterval includes the start LocalTime and excludes the end LocalTime. However, an TimeInterval
+    /// may be missing its start or end, in which case the TimeInterval is deemed to be infinite in that
+    /// direction.
+    /// </para>
+    /// <para>
+    /// The end may equal the start (resulting in an empty TimeInterval),
+    /// but it can also be before the start: in this case property IsEndInNextDay will be true
+    /// </para>
+    /// </remarks>
+    /// <threadsafety>This type is an immutable value type. See the thread safety section of the user guide for more information.</threadsafety>
+    public readonly struct TimeInterval : IEquatable<TimeInterval>
+    {
+        /// <summary>The start of the interval.</summary>
+        private readonly LocalTime? start;
+
+        /// <summary>The end of the interval. It can be less than start.</summary>
+        private readonly LocalTime? end;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TimeInterval"/> struct.
+        /// The TimeInterval includes the <paramref name="start"/> LocalTime and excludes the
+        /// <paramref name="end"/> LocalTime. The end may equal the start (resulting in an empty TimeInterval).
+        /// </summary>
+        /// <param name="start">The start <see cref="LocalTime"/>.</param>
+        /// <param name="end">The end <see cref="LocalTime"/>.</param>
+        public TimeInterval(LocalTime start, LocalTime end)
+        {
+            this.start = start;
+            this.end = end;
+            IsEndInNextDay = start > end;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TimeInterval"/> struct from two nullable <see cref="LocalTime"/>
+        /// values.
+        /// </summary>
+        /// <remarks>
+        /// If the start is null, the TimeInterval is deemed to stretch to the start of time. If the end is null,
```

Comment:
I'd use "start of the day" and "end of the day" rather than "start of time" and "end of time".